pkg/cloud/services/compute/instance_types.go

```diff
@@ -132,51 +135,100 @@ func (is *InstanceStatus) InstanceIdentifier() *InstanceIdentifier {
 
 // NetworkStatus returns an InstanceNetworkStatus object for an InstanceStatus.
 func (is *InstanceStatus) NetworkStatus() (*InstanceNetworkStatus, error) {
-	addresses := make(map[string][]networkInterface)
-
+	// Gophercloud doesn't give us a struct for server addresses: we get a
+	// map of networkname -> interface{}. That interface{} is a list of
+	// addresses as in the example output here:
+	// https://docs.openstack.org/api-ref/compute/?expanded=show-server-details-detail#show-server-details
+	//
+	// Here we convert the interface{} into something more usable by
+	// marshalling it to json, then unmarshalling it back into our own
+	// struct.
+	addressesByNetwork := make(map[string][]corev1.NodeAddress)
 	for networkName, b := range is.server.Addresses {
 		list, err := json.Marshal(b)
 		if err != nil {
 			return nil, fmt.Errorf("error marshalling addresses for instance %s: %w", is.ID(), err)
 		}
-		var networkList []networkInterface
-		err = json.Unmarshal(list, &networkList)
+		var interfaceList []networkInterface
+		err = json.Unmarshal(list, &interfaceList)
 		if err != nil {
 			return nil, fmt.Errorf("error unmarshalling addresses for instance %s: %w", is.ID(), err)
 		}
 
-		addresses[networkName] = networkList
-	}
+		var addresses []corev1.NodeAddress
+		for i := range interfaceList {
+			address := &interfaceList[i]
 
-	return &InstanceNetworkStatus{addresses}, nil
-}
+			// Only consider IPv4
+			if address.Version != 4 {
+				continue
+			}
 
-func (ns *InstanceNetworkStatus) IP() string {
-	// Return the last listed non-floating IPv4 from the last listed network
-	// This behaviour is wrong, but consistent with the previous behaviour
-	// https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/4debc1fc4742e483302b0c36b16c076977bd165d/pkg/cloud/services/compute/instance.go#L973-L998
-	// XXX: Fix this
-	for _, vifs := range ns.addresses {
-		for _, vif := range vifs {
-			if vif.Version == 4.0 && vif.Type != "floating" {
-				return vif.Address
+			var addressType corev1.NodeAddressType
+			switch address.Type {
+			case "floating":
+				addressType = corev1.NodeExternalIP
+			case "fixed":
+				addressType = corev1.NodeInternalIP
+			default:
+				// Ignore unknown address types
```

Comment:
maybe add some log as reference? though should not reach here..

Reply:
I thought about this, and I'm not sure.

The most likely reason that we would get an unknown address type here is that OpenStack introduced a new address type which we don't use, yet. This would be a new feature, and logging something here because of a new feature we don't support yet would be very noisy. I guess we could log this and IPv6 addresses at debug level just in case?

Interestingly, I was speaking to a Nova developer about `addresses` yesterday, and his impression is that `addresses` is more likely to go away entirely rather than gain new features because it's viewed as a proxy API. The replacement would be to directly query neutron for ports associated with the device id. The only reason I didn't make that change here is that as long as Nova is caching NetworkInfo, getting it from Nova requires 1 less API call.

TL;DR I might add this at debug.

Reply:
I've added this.